deepchem/data/data_loader.py

```diff
@@ -968,42 +968,55 @@ def create_dataset(self,
     if isinstance(input_files, str):
       input_files = [input_files]
 
-    def shard_generator():  # TODO Enable sharding with shard size parameter
+    def shard_generator():
       for input_file in input_files:
-        if self.legacy:
+        shards = _read_file(input_file)
+        if self.legacy:  # Using legacy logic
           X = encode_bio_sequence(input_file)
-        else:
-          sequences = _read_file(input_file)
-          X = self.featurizer(sequences)
-        ids = np.ones(len(X))
-        # (X, y, w, ids)
-        yield X, None, None, ids
+          ids = np.ones(len(X))
+          # (X, y, w, ids)
+          yield X, None, None, ids
+        else:  # Not using legacy logic
+          for shard in shards:
+            X = self.featurizer(shard)
+            ids = np.ones(len(X))
```

Comment:
A better default here is probably to do an index. That is, 1,2,3,etc. You can do this by maintaining a global count and doing something like
```
ids = count + np.arange(len(X))
```